### PR TITLE
feat(rpc): Implement `state_getReadProof` rpc call

### LINKED ISSUE #1766: Implement `state_getReadProof` RPC call
## Task summary

- Implement `state_getReadProof` RPC call

## Specification reference

- [Spec reference](https://github.com/w3f/PSPs/blob/master/PSPs/drafts/psp-6.md#11112-state_getreadproof)

## Other information and links

- N/A

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

- Update the `Trie.GenerateProof` to handle more than one storage key, and use a `map` to avoid duplicate node hashes
- Create the `core.GetReadProofAt` which receives a block hash and the storage keys and returns the proof

## Tests

```
go test -run ^TestGetReadProofAt$ github.com/ChainSafe/gossamer/dot/core
go test -run ^TestGetReadProof_WhenCoreAPIReturnsError$ github.com/ChainSafe/gossamer/dot/rpc/modules
go test -run ^TestGetReadProof_WhenReturnsProof$ github.com/ChainSafe/gossamer/dot/rpc/modules
go test -run ^TestProofGeneration$ github.com/ChainSafe/gossamer/lib/trie
```

## Issues

- Closes #1766 

## Primary Reviewer 

- @noot

### CODE REVIEW

File: dot/rpc/modules/state.go [hunk 1]
```diff
@@ -279,6 +291,37 @@ func (sm *StateModule) GetMetadata(r *http.Request, req *StateRuntimeMetadataQue
...
+	for i, hexKey := range req.Keys {
+		bKey, err := common.HexToBytes(hexKey)
+		if err != nil {
+			return err
+		}
+
+		keys[i] = bKey
+	}
+
+	fmt.Println("GET READ PROOF RPC", req.Keys, len(keys))
```
Comment (noot) [inline dot/rpc/modules/state.go:306]: remove

File: dot/state/block.go [hunk 2]
```diff
@@ -734,3 +745,8 @@ func (bs *BlockState) StoreRuntime(hash common.Hash, rt runtime.Instance) {
 func (bs *BlockState) GetNonFinalisedBlocks() []common.Hash {
 	return bs.bt.GetAllBlocks()
 }
+
+// GenerateTrieProof returns the proofs related to the keys on the state root trie
+func (bs *BlockState) GenerateTrieProof(stateRoot common.Hash, keys [][]byte) ([][]byte, error) {
+	return trie.GenerateProof(stateRoot[:], keys, bs.db)
+}
```
Comment (noot) [inline dot/state/block.go:752]: this would make more sense in the `storageState`

File: lib/trie/proof.go [hunk 3]
```diff
@@ -19,126 +19,274 @@ package trie
...
+		unwindStack(&stack, proofs, nk)
 
-		currNode, err := decodeBytes(enc)
+		lookup := NewLookup(root, db)
+		recorder := new(Recorder)
+		expectedValue, err := lookup.Find(nk, recorder)
 		if err != nil {
-			return false, fmt.Errorf("could not decode node bytes: %w", err)
+			return nil, err
+		}
```
Comment (noot) [inline lib/trie/proof.go:153]: the value isn't required for the proof. I'm also not sure what the purpose of the code after here is, as when I tried it previously the proof was already in the `recorder` at this point

File: dot/rpc/modules/state.go (hunk 1 above)
Comment (EclesioMeloJunior) [reply dot/rpc/modules/state.go]: Removed

File: dot/state/block.go (hunk 2 above)
Comment (EclesioMeloJunior) [reply dot/state/block.go]: Updated

File: lib/trie/lookup.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+// Lookup struct holds the state root and database reference
+// used to retrieve trie information from database
+type Lookup struct {
+	// root to start the lookup
+	root []byte
+	db   chaindb.Database
+}
+
+// NewLookup returns a Lookup to helps the proof generator
+func NewLookup(h []byte, db chaindb.Database) *Lookup {
```
[the remark this answers is not in the document]
Comment (EclesioMeloJunior) [reply lib/trie/lookup.go]: changed to `rootHash`

File: lib/trie/proof.go (hunk 3 above)
Comment (EclesioMeloJunior) [reply lib/trie/proof.go]: you're right, the proof is inside recorder, I remove the unecessary code and I added a map of `trackedProofs` to dont duplicate already added proofs

File: lib/trie/lookup.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
+package trie
```
Comment (jimjbrettj) [inline lib/trie/lookup.go:1]: Does the lib/.DS_Store need to be in this PR?

File: lib/trie/lookup.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import (
+	"bytes"
+	"errors"
+
+	"github.com/ChainSafe/chaindb"
+)
+
+var (
+	// ErrProofNodeNotFound when a needed proof node is not in the database
+	ErrProofNodeNotFound = errors.New("cannot found the state root on storage")
```
Comment (jimjbrettj) [inline lib/trie/lookup.go:12]: nit: maybe should be "in storage" instead of  "on storage"

File: lib/trie/proof.go [hunk 7]
```diff
@@ -28,117 +27,41 @@ import (
...
-		var (
-			hashNode    []byte
-			encHashNode []byte
-			err         error
-		)
-
-		if encHashNode, hashNode, err = n.encodeAndHash(); err != nil {
-			return 0, fmt.Errorf("problems while encoding and hashing the node: %w", err)
-		}
+		fmt.Println("Len of records ", len(*recorder))
```
Comment (jimjbrettj) [inline lib/trie/proof.go:48]: I think should remove comment or make a log

File: lib/trie/proof.go [hunk 8]
```diff
@@ -28,117 +27,41 @@ import (
...
-}
+	proofs := make([][]byte, 0)
 
-// VerifyProof checks merkle proofs given an proof
-func VerifyProof(rootHash common.Hash, key []byte, db chaindb.Reader) (bool, error) {
-	key = keyToNibbles(key)
-	if len(key) == 0 {
-		return false, ErrEmptyNibbles
+	for _, p := range trackedProofs {
+		fmt.Printf("tracked proofs: 0x%x\n", p)
```
Comment (jimjbrettj) [inline lib/trie/proof.go:62]: Should this also use logger?

File: lib/trie/proof_test.go [hunk 9]
```diff
@@ -17,237 +17,39 @@
...
-
-func modifyLastBytes(b []byte) []byte {
-	newB := make([]byte, len(b))
-	copy(newB[:], b)
-
-	rb := randBytes(12)
-	copy(newB[20:], rb)
-
-	return newB
+	// TODO: use the verify_proof function to assert the tests
```
Comment (jimjbrettj) [inline lib/trie/proof_test.go:53]: Is there plans to do this in a different PR?

File: lib/trie/recorder.go [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+// Peek returns the current item the cursor is on but dont increment the cursor by 1
+func (r *Recorder) Peek() *NodeRecord {
+	if r.HasNext() {
+		return &(*r)[0]
+	}
+	return nil
+}
+
+// HasNext returns bool if there is data inside the slice
+func (r *Recorder) HasNext() bool {
```
Comment (jimjbrettj) [inline lib/trie/recorder.go:38]: Nit: Would IsEmpty be a more appropriate name for this function? HasNext makes me think more of traversing through a linked list. Maybe thats just me thought, just a suggestion

**Review by jimjbrettj — APPROVED** [review]
Small suggestions and comments, overall looks great!

File: lib/trie/lookup.go (hunk 5 above)
Comment (EclesioMeloJunior) [reply lib/trie/lookup.go:1]: removed!

File: lib/trie/lookup.go (hunk 6 above)
Comment (EclesioMeloJunior) [reply lib/trie/lookup.go:12]: updated!

File: lib/trie/proof.go (hunk 7 above)
Comment (EclesioMeloJunior) [reply lib/trie/proof.go:48]: removed prints!

File: lib/trie/proof.go (hunk 8 above)
Comment (EclesioMeloJunior) [reply lib/trie/proof.go:62]: removed prints!

File: lib/trie/proof_test.go (hunk 9 above)
Comment (EclesioMeloJunior) [reply lib/trie/proof_test.go:53]: yes! the issue to create the `verify_proof` function is already done!

File: lib/trie/recorder.go (hunk 10 above)
Comment (EclesioMeloJunior) [reply lib/trie/recorder.go:38]: makes sense! I've updated the func name

File: lib/trie/lookup.go (hunk 5 above)
Comment (noot) [reply lib/trie/lookup.go:1]: add license pls

File: lib/trie/lookup.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+import (
+	"bytes"
+	"errors"
+
+	"github.com/ChainSafe/chaindb"
+)
+
+var (
+	// ErrProofNodeNotFound when a needed proof node is not in the database
+	ErrProofNodeNotFound = errors.New("cannot found the state root in storage")
```
Comment (noot) [inline lib/trie/lookup.go:12]:
```suggestion
	ErrProofNodeNotFound = errors.New("cannot find a trie node in storage")
```

File: lib/trie/lookup.go [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+// Lookup struct holds the state root and database reference
+// used to retrieve trie information from database
+type Lookup struct {
```
Comment (noot) [inline lib/trie/lookup.go:17]: the types/funcs in this file can be unexported, as they aren't used by any other package

File: lib/trie/proof.go [hunk 13]
```diff
@@ -28,117 +26,38 @@ import (
...
-
-			if bytes.Equal(n.key, key) {
-				break proveLoop
-			}
-
-			return 0, errors.New("leaf node doest not match the key")
+// GenerateProof receive the keys to proof, the trie root and a reference to database
+// will
+func GenerateProof(root []byte, keys [][]byte, db chaindb.Database) ([][]byte, error) {
+	trackedProofs := make(map[string][]byte)
```
Comment (noot) [inline lib/trie/proof.go:34]: does this work with multiple keys at the moment?

File: lib/trie/recorder.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
+package trie
```
Comment (noot) [inline lib/trie/recorder.go:1]: add license

File: lib/trie/recorder.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+// NodeRecord represets a record of a visited node
+type NodeRecord struct {
```
Comment (noot) [inline lib/trie/recorder.go:4]: don't think the types/funcs in here need to be exported either

File: lib/trie/test_utils.go [hunk 16]
```diff
@@ -71,3 +71,31 @@ func generateRandomTest(t testing.TB, kv map[string][]byte) Test {
...
 }
+
+// KV helps to export the entries of the generated keys
+type KV struct {
+	K []byte
+	V []byte
+}
+
+// RandomTrieTest generate a trie with random entries
+func RandomTrieTest(t *testing.T, n int) (*Trie, map[string]*KV) {
```
Comment (noot) [inline lib/trie/test_utils.go:82]: same here

**Review by noot — COMMENTED** [review]
overall looks great, just a few comments!

File: lib/trie/proof.go (hunk 13 above)
Comment (EclesioMeloJunior) [reply lib/trie/proof.go:34]: yep!

File: lib/trie/test_utils.go (hunk 16 above)
Comment (EclesioMeloJunior) [reply lib/trie/test_utils.go:82]: Updated!

File: lib/trie/recorder.go (hunk 15 above)
Comment (EclesioMeloJunior) [reply lib/trie/recorder.go]: Updated!

File: lib/trie/recorder.go (hunk 14 above)
Comment (EclesioMeloJunior) [reply lib/trie/recorder.go:1]: Added!

File: dot/rpc/modules/state.go [hunk 17]
```diff
@@ -28,6 +28,12 @@ import (
 	"github.com/ChainSafe/gossamer/pkg/scale"
 )
 
+//StateGetReadProofRequest json fields
+type StateGetReadProofRequest struct {
+	Keys []string
+	Hash common.Hash
```
Comment (noot) [inline dot/rpc/modules/state.go:34]: should this be `*common.Hash` if it's optional?

File: lib/trie/lookup.go (hunk 5 above)
Comment (noot) [reply lib/trie/lookup.go:1]: ^ poke on the above

File: lib/trie/proof.go [hunk 18]
```diff
@@ -28,117 +26,38 @@ import (
...
-		case *leaf:
-			nodes = append(nodes, n)
-
-			if bytes.Equal(n.key, key) {
-				break proveLoop
-			}
-
-			return 0, errors.New("leaf node doest not match the key")
+// GenerateProof receive the keys to proof, the trie root and a reference to database
+// will
```
Comment (noot) [inline lib/trie/proof.go:32]: remove this?

File: lib/trie/proof.go (hunk 13 above)
Comment (noot) [reply lib/trie/proof.go:34]: what if multiple node proofs overlap? does this return the same as proof substrate? I'm just trying to understand how this works for multiple proofs within one map (since there's no ordering)

**Review by noot — APPROVED** [review]
overall looks good, just a couple questions/comments
